internal/internal_task_pollers.go

```diff
@@ -303,6 +307,36 @@ func newWorkflowTaskPoller(
 	}
 }
 
+// Best-effort attempt to indicate to Matching service that this workflow task
+// poller's sticky queue will no longer be polled. Should be called when the
+// poller is stopping. Failure to call ShutdownWorker is logged, but otherwise
+// ignored.
+func (wtp *workflowTaskPoller) Cleanup() error {
+	ctx := context.Background()
+	grpcCtx, cancel := newGRPCContext(ctx, grpcMetricsHandler(wtp.metricsHandler))
+	defer cancel()
+
+	_, err := wtp.service.ShutdownWorker(grpcCtx, &workflowservice.ShutdownWorkerRequest{
```

Comment:
> Is there any way for us to test in our integration tests that this works and has effect? Even if not now, can we open an issue for when it is available server side to confirm it has effect (e.g. that sticky was reset across two workers)? 

Not exactly. The call signals Matching to route tasks away from a shutdown worker's sticky queue (where the total timeout is 10 seconds). However, the caller that would actually see the effect from the rerouting isn't the end client, it's History service, which [handles the rerouting](https://github.com/temporalio/temporal/blob/99f15c4dabf9a22763aaa13653c83a832bd0fffa/service/history/transfer_queue_active_task_executor.go#L272-L287). The end result is that any future tasks previously destined for the sticky queue will have a lower schedule-to-start latency.

The important takeaway is that this isn't very observable from a testing point of view this far away in the stack - the signals between success and failure are fuzzy, it's just latency. If the call fails, the worst case scenario is that schedule-to-start is more latent (by up to 10s). If the call succeeds, we know schedule-to-start is hopefully less latent, but that isn't guaranteed by any means. `ShutdownWorker` could work perfectly, and the task's `ScheduleToStart` may be slow from an entirely unrelated reason in History or Matching services. 

> My fear is that we may do something wrong on the request building or something and this gRPC call always errors and we would just never know that it never worked.

Because the effect of this call takes place entirely within the service, asynchronously, IMO the testing needs to verify two independent facts:

- Does the worker call `ShutdownWorker` and receive positive acknowledgement from the service? The call doesn't have an effect directly visible to the worker, so here we instead need to verify that the server will validate and accept the request.
- Does the service process `ShutdownWorker` properly/does the call have an effect? Because we're testing deeper within the service, we get a strong discrete signal for if the call had an effect; I can directly call `MatchingService` and assert that it returns `StickyWorkerUnavailable` for a shut down queue. This is a much better signal for test success than merely trying to guess based on downstream latency.

Reply:
:+1: I wonder if we can demonstrate a user-visible effect, e.g. such as having two workers, and the worker that got a sticky workflow is able to be shutdown and have the next workflow task immediately go to the next worker without penalty.

Regardless, I understand we cannot assert this behavior in integration tests yet anyways.

Reply:
@lina-temporal did you add any test to verify `ShutdownWoker` is only treated as best effort? I was expecting to see a unit test that mocked the service stub returned an error for `ShutdownWoker` and verify that from the users perspective the shutdown still succeeded. If I missed it please just point me at the test.

Reply:
Sure, added a test to assert that `Cleanup` is best effort (`TestCleanupIsBestEffort`).